### PR TITLE
Subnet Identities

### LINKED ISSUE #708: Subnet Identity
## Description

Currently, the subnet registration process doesn't include identity information, and the `SubnetInfo` struct lacks fields for storing this data. We need to enhance the subnet registration process to include identity information and extend the `SubnetInfo` struct to accommodate these new fields.

This enhancement will allow subnet owners to provide more detailed information about their subnets, making it easier for users and administrators to identify and interact with different subnets in the Bittensor network.

## Acceptance Criteria

1. Extend the `SubnetInfo` struct in `pallets/subtensor/src/rpc_info/subnet_info.rs` to include the following new fields:
   - `subnet_name: Option<Vec<u8>>`
   - `github_repo: Option<Vec<u8>>`
   - `subnet_contact: Option<Vec<u8>>`

2. Modify the subnet registration process to accept and store the new identity information.

3. Update the `get_subnet_info` function to retrieve and return the new identity fields.

4. Ensure that the subnet deregistration process clears the identity information.

5. Add appropriate error handling for cases where required identity information is missing during registration.

## Tasks

- [ ] Update the `SubnetInfo` struct in `pallets/subtensor/src/rpc_info/subnet_info.rs`.
- [ ] Modify the `register_subnet` function to accept and store identity information.
- [ ] Update the `get_subnet_info` function to include the new identity fields in its return value.
- [ ] Modify the `deregister_subnet` function to clear identity information.
- [ ] Add new error types for missing required identity information.
- [ ] Update any relevant tests to cover the new functionality.
- [ ] Update documentation to reflect the changes in subnet registration and information retrieval.

## Additional Considerations

- Consider making `subnet_name` and `github_repo` required fields, while keeping `subnet_contact` optional.
- Evaluate the impact of these changes on existing subnets and consider a migration strategy if necessary.
- Consider adding a function to update subnet identity information separately from the registration process.

### PR DESCRIPTION
**Pull request by JohnReedV**
## Description
Adds subnet Identities

Done: 
- Define subnet identity structs and storage items.
- Add extrinsic ` set_subnet_identity`.
- Add the subnet identity fields to `SubnetInfo`.
- Remove the identity when the subnet is dissolved. 
- Write unit tests
- Integrate identities into subnet registration process 

Todo:
- Test Manually

## Related Issue(s)

- Closes #708

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Other (please describe):

## Breaking Change
- The extrinsic `user_add_network` now takes a second parameter `identity: Option<SubnetIdentityOf>`
- `SubnetInfo` has a new parameter `identity: Option<SubnetIdentity>`.

### CODE REVIEW

File: pallets/subtensor/src/lib.rs [hunk 1]
```diff
@@ -152,6 +152,18 @@ pub mod pallet {
         pub additional: Vec<u8>,
     }
 
+    ///  Struct for Prometheus.
+    pub type SubnetIdentityOf = SubnetIdentity;
+    /// Data structure for Prometheus information.
+    #[derive(Encode, Decode, Default, TypeInfo, Clone, PartialEq, Eq, Debug)]
+    pub struct SubnetIdentity {
```
Comment (open-junius) [inline pallets/subtensor/src/lib.rs:177]: we need add freeze_struct for the struct.

File: pallets/subtensor/src/coinbase/root.rs [hunk 2]
```diff
@@ -894,17 +894,24 @@ impl<T: Config> Pallet<T> {
...
     ///
     /// # Raises:
     /// * 'TxRateLimitExceeded': If the rate limit for network registration is exceeded.
     /// * 'NotEnoughBalanceToStake': If there isn't enough balance to stake for network registration.
     /// * 'BalanceWithdrawalError': If an error occurs during balance withdrawal for network registration.
     ///
-    pub fn user_add_network(origin: T::RuntimeOrigin) -> dispatch::DispatchResult {
+    pub fn user_add_network(
+        origin: T::RuntimeOrigin,
+        identity: Option<SubnetIdentityOf>,
```
Comment (distributedstatemachine) [inline pallets/subtensor/src/coinbase/root.rs:913]: Why have you decided on making idenitity optional ?

File: pallets/subtensor/src/lib.rs [hunk 3]
```diff
@@ -60,7 +60,7 @@ extern crate alloc;
 #[frame_support::pallet]
 pub mod pallet {
 
-    use crate::migrations;
+    use crate::{freeze_struct, migrations};
```
Comment (distributedstatemachine) [inline pallets/subtensor/src/lib.rs]: how do this line change affect the task ?

Comment (JohnReedV) [reply pallets/subtensor/src/lib.rs]: Importing `freeze_struct` to freeze the `SubnetIdentity` struct.

File: pallets/subtensor/src/coinbase/root.rs (hunk 2 above)
Comment (JohnReedV) [reply pallets/subtensor/src/coinbase/root.rs:913]: Mostly to follow the existing pattern set by delegate identities. I don't have a strong preference here.

**Review by distributedstatemachine — APPROVED** [review]

**Review by distributedstatemachine — DISMISSED** [review]

**Review by unconst — DISMISSED** [review]

**Pull request merged**